### PR TITLE
Fixed #33966 -- Added support for using KeyTextTransform from lookup.

### PR DESCRIPTION
**Pull request by AllenJonathan**

### CODE REVIEW

File: django/db/models/fields/json.py [hunk 1]
```diff
@@ -379,6 +380,20 @@ def as_mysql(self, compiler, connection):
...
 
+    @classmethod
+    def from_lookup(cls, lookup):
+        transform, *keys = lookup.split(LOOKUP_SEP)
+        if not keys:
+            print("decide")
+        for key in keys:
+            print(key, transform)
+            transform = cls(key, transform)
+            print(transform)
```
Comment (felixxm) [inline django/db/models/fields/json.py:391]: Please remove debug  `print`'s.

File: django/db/models/fields/json.py [hunk 2]
```diff
@@ -379,6 +380,20 @@ def as_mysql(self, compiler, connection):
             json_path = compile_json_path(key_transforms)
             return "(%s ->> %%s)" % lhs, tuple(params) + (json_path,)
 
+    @classmethod
+    def from_lookup(cls, lookup):
+        transform, *keys = lookup.split(LOOKUP_SEP)
+        if not keys:
+            print("decide")
```
Comment (felixxm) [inline django/db/models/fields/json.py:387]: This should raise a `ValueError`.

**Review by felixxm — COMMENTED** [review]
@AllenJonathan Thanks :+1:

File: django/db/models/fields/json.py [hunk 3]
```diff
@@ -379,6 +380,18 @@ def as_mysql(self, compiler, connection):
             json_path = compile_json_path(key_transforms)
             return "(%s ->> %%s)" % lhs, tuple(params) + (json_path,)
 
+    @classmethod
+    def from_lookup(cls, lookup):
+        transform, *keys = lookup.split(LOOKUP_SEP)
+        if not keys:
+            raise ValueError("Missing keys")
```
Comment (felixxm) [inline django/db/models/fields/json.py:387]: Please add test for this error.

File: docs/topics/db/queries.txt [hunk 4]
```diff
@@ -1050,6 +1050,20 @@ To query for missing keys, use the ``isnull`` lookup::
     >>> Dog.objects.filter(data__owner__isnull=True)
     <QuerySet [<Dog: Shep>]>
 
+.. class:: json.KT(lookup)
```
Comment (felixxm) [inline docs/topics/db/queries.txt:1053]:
```suggestion
``KT()`` expressions
~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 4.2

.. class:: json.KT(lookup)
```

File: docs/topics/db/queries.txt [hunk 5]
```diff
@@ -1050,6 +1050,20 @@ To query for missing keys, use the ``isnull`` lookup::
...
+
+    A class that returns the resulting transformed JSON in the form of text.
+    ``lookup`` argument be a string starting with the name of the
+    :class:`~django.db.models.JSONField` and subsequent key or index transforms
+    separated by ``__`` (double underscore).
```
Comment (felixxm) [inline docs/topics/db/queries.txt:1058]: Maybe:
```suggestion
Represents the text value of a key, index, or path transform of
:class:`~django.db.models.JSONField`. You can use the double underscore
notation in `lookup` to chain dictionary keys and index transforms. For
example::
```

File: docs/topics/db/queries.txt [hunk 6]
```diff
@@ -1050,6 +1050,20 @@ To query for missing keys, use the ``isnull`` lookup::
...
+
+    To text transform :class:`~django.db.models.JSONField` - ``Dog.data``::
+
+      >>> Dog.objects.create(name='Shep', data={'breed': ['collie', 'lhasa apso']})
+      <Dog: Shep>
+      >>> Dogs.objects.annotate(breed=KT("data__breed__1")).filter(breed='lhasa apso')
+      <QuerySet [<Dog: Shep>]>
```
Comment (felixxm) [inline docs/topics/db/queries.txt:1065]: I would add chain `KT` with a lookup and a second annotation, e.g.
```suggestion
    >>> Dog.objects.create(name="Shep", data={
    ...     "owner": {"name": "Bob"},
    ...     "breed": ["collie", "lhasa apso"],
    ... })
    <Dog: Shep>
    >>> Dogs.objects.annotate(
    ...     first_breed=KT("data__breed__1"),
    ...     owner=KT("data__owner__name"),
    ... ).filter(breed__startswith="lhasa", owner="Bob")
    <QuerySet [<Dog: Shep>]>

File: django/db/models/fields/json.py (hunk 3 above)
Comment (felixxm) [inline django/db/models/fields/json.py:387]:
```suggestion
            raise ValueError("lookup cannot be empty.")
```

File: tests/model_fields/test_jsonfield.py [hunk 7]
```diff
@@ -1078,3 +1079,18 @@ def test_join_key_transform_annotation_expression(self):
             ).filter(chain=F("related_key__0")),
             [related_obj],
         )
+
+    def test_key_text_transform_from_lookup(self):
+        qs1 = NullableJSONModel.objects.annotate(b=KT("value__bax__foo")).filter(
+            b__contains="ar"
+        )
+        self.assertCountEqual(qs1, [self.objs[7]])
```
Comment (felixxm) [inline tests/model_fields/test_jsonfield.py:1087]:
```suggestion
        self.assertSequenceEqual(qs1, [self.objs[7]])
```

File: tests/model_fields/test_jsonfield.py [hunk 8]
```diff
@@ -1078,3 +1079,18 @@ def test_join_key_transform_annotation_expression(self):
...
+        qs2 = NullableJSONModel.objects.annotate(c=KT("value__o")).filter(
+            c__contains="uot"
+        )
+        self.assertCountEqual(qs2, [self.objs[4]])
```
Comment (felixxm) [inline tests/model_fields/test_jsonfield.py:1091]:
```suggestion
        self.assertSequenceEqual(qs2, [self.objs[4]])
```

**Review by felixxm — COMMENTED** [review]
@AllenJonathan Thanks :+1: I left comments, please add also a release note.

File: django/db/models/fields/json.py (hunk 3 above)
Comment (AllenJonathan) [reply django/db/models/fields/json.py:387]: Provide a non-empty `lookup` - giving only `transform` would produce this error. Wouldn't this be incorrect?

**Comment by AllenJonathan** [comment]
Just the error needs discussion. I've made all the other suggested changes.

File: django/db/models/fields/json.py (hunk 3 above)
Comment (felixxm) [reply django/db/models/fields/json.py:387]: Right, maybe _"lookup must contains key or index transforms."_ :thinking: .

**Comment by felixxm** [comment]
@AllenJonathan Thanks :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Pull request merged**